src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs

```diff
@@ -51,105 +55,129 @@ protected override Task DisposeConnection()
-            var clientContext = new ClientContext();
-            try
-            {
-                await clientContext.Output.WriteAsync(openConnectionMessage);
-                if(!_clientConnections.TryAdd(connectionId, clientContext))
-                {
-                    Log.DuplicateConnectionId(Logger, connectionId, null);
-                    throw new ArgumentException("ConnectionId already exists.");
-                }
+            var clientContext = new ClientContext(connectionId);
 
-                // Writing from the application to the service
-                _ = ProcessMessageAsync(connectionId);
+            if (_clientConnections.TryAdd(connectionId, clientContext) &&
+                _clientConnectionManager.TryAdd(connectionId, this))
+            {
+                clientContext.ApplicationTask = ProcessMessageAsync(clientContext, clientContext.CancellationToken);
+                return ForwardMessageToApplication(connectionId, openConnectionMessage);
             }
-            catch (Exception e)
+            else
             {
-                // Fail to write initial open connection message to channel
-                Log.ConnectedStartingFailed(Logger, connectionId, e);
-                // Close channel and notify client to close connection
-                clientContext.Output.TryComplete();
-                await WriteAsync(new CloseConnectionMessage(connectionId, e.Message));
+                Log.DuplicateConnectionId(Logger, connectionId, null);
+                return Task.CompletedTask;
             }
         }
 
-        protected override async Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
+        protected override Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
+        {
+            return ForwardMessageToApplication(closeConnectionMessage.ConnectionId, closeConnectionMessage);
+        }
+
+        protected override Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        {
+            return ForwardMessageToApplication(connectionDataMessage.ConnectionId, connectionDataMessage);
+        }
+
+        private async Task ForwardMessageToApplication(string connectionId, ServiceMessage message)
         {
-            var connectionId = closeConnectionMessage.ConnectionId;
             if (_clientConnections.TryGetValue(connectionId, out var clientContext))
             {
                 try
                 {
-                    await clientContext.Output.WriteAsync(closeConnectionMessage);
+                    await clientContext.Output.WriteAsync(message);
                 }
                 catch (Exception e)
                 {
-                    Log.FailToWriteMessageToApplication(Logger, connectionId, e);
+                    Log.FailToWriteMessageToApplication(Logger, message.GetType().Name, connectionId, e);
+                    await PerformDisconnectCore(connectionId, true);
```

Comment:
>await PerformDisconnectCore(connectionId, true); [](start = 20, length = 48)

Seems only under the case which probably like channel broken or overflow that need to wait application, then will this be necessary to wait under this case? Normally wait won't get any better.